relayer/provider/cosmos/provider.go

```diff
@@ -888,16 +888,89 @@ func (cc *CosmosProvider) MsgTransfer(amount sdk.Coin, dstChainId, dstAddr, srcP
 // MsgRelayTimeout constructs the MsgTimeout which is to be sent to the sending chain.
 // The counterparty represents the receiving chain where the receipts would have been
 // stored.
-func (cc *CosmosProvider) MsgRelayTimeout(ctx context.Context, dst provider.ChainProvider, dsth int64, packet provider.RelayPacket, dstChanId, dstPortId, srcChanId, srcPortId string) (provider.RelayerMessage, error) {
+func (cc *CosmosProvider) MsgRelayTimeout(
+	ctx context.Context,
+	dst provider.ChainProvider,
+	dsth int64,
+	packet provider.RelayPacket,
+	dstChanId, dstPortId, srcChanId, srcPortId string,
+	order chantypes.Order,
+) (provider.RelayerMessage, error) {
 	var (
 		acc string
 		err error
+		msg provider.RelayerMessage
 	)
 	if acc, err = cc.Address(); err != nil {
 		return nil, err
 	}
 
+	switch order {
+	case chantypes.UNORDERED:
+		msg, err = cc.unorderedChannelTimeoutMsg(ctx, dst, dsth, packet, acc, dstChanId, dstPortId, srcChanId, srcPortId)
+		if err != nil {
+			return nil, err
+		}
+	case chantypes.ORDERED:
+		msg, err = cc.orderedChannelTimeoutMsg(ctx, dst, dsth, packet, acc, dstChanId, dstPortId, srcChanId, srcPortId)
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, fmt.Errorf("invalid order type %s, order should be %s or %s",
+			order.String(), chantypes.ORDERED.String(), chantypes.UNORDERED.String())
+	}
+
+	return msg, nil
+}
+
+func (cc *CosmosProvider) orderedChannelTimeoutMsg(
+	ctx context.Context,
+	dst provider.ChainProvider,
+	dsth int64,
+	packet provider.RelayPacket,
+	acc, dstChanId, dstPortId, srcChanId, srcPortId string,
+) (provider.RelayerMessage, error) {
+	seqRes, err := dst.QueryNextSeqRecv(ctx, dsth, dstChanId, dstPortId)
+
+	switch {
+	case err != nil:
+		return nil, err
+	case seqRes.Proof == nil:
+		return nil, fmt.Errorf("timeout packet next sequence received proof seq(%d) is nil", packet.Seq())
+	case seqRes == nil:
```

Comment:
Expression switches are evaluated top to bottom, so checking `seqRes.Proof == nil` will panic if `seqRes == nil`.

I would do away with the switch altogether and just have three if statements in a row, followed by the unconditional case.

Reply:
Ahh good call, just did a quick copy/paste from how we were doing this before 🤦 
I switched to just using if statements like you suggested and it looks much better, I'll go ahead and do this for the other timeout function as well.